In [1]:
# Import useful libraries
import time
import config
import numpy as np
import pandas as pd
import os
import re
import requests
import tweepy
import json
from PIL import Image
from io import BytesIO
from tabulate import tabulate

In [2]:
class Color:
    purple = '\033[95m'
    cyan = '\033[96m'
    darkcyan = '\033[36m'
    blue = '\033[94m'
    green = '\033[92m'
    yellow = '\033[93m'
    red = '\033[91m'
    bold = '\033[1m'
    underline = '\033[4m'
    end = '\033[0m'

## Gather and Read Data
---

In [3]:
# Read the twitter archive data provided
wrd_archive = pd.read_csv('./twitter-archive-enhanced.csv')
wrd_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [4]:
# Programmatically download the image predictions
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
file_name = url.split('/')[-1]
response = requests.get(url)

start = time.time()
with open(file_name, 'wb') as f:
    f.write(response.content)
    
print('Process completed in {} seconds'.format(time.time()-start))

Process completed in 0.003099203109741211 seconds


In [5]:
# Read in the image predictions
img_predictions = pd.read_csv('./image-predictions.tsv', sep='\t')
img_predictions.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

In [6]:
# Create an API object to gater twitter data
consumer_key = config.API_KEY
consumer_secret = config.API_KEY_SECRET
access_token = config.ACCESS_TOKEN
access_secret = config.ACCESS_TOKEN_SECRET

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit =True,
                wait_on_rate_limit_notify=True)

In [7]:
if False:
# Pull tweet information using the ids in wrd_archive
# Extract the tweet ids from the wrd dataframe
    tweet_ids = wrd_archive['tweet_id']
    success, failure, counter = (0, 0, 0)
    failed_attempts = {}
    print('\033[1m'+'COMMENCING JSON EXTRACTION TASK'+'\033[0m'+'\n'+'-'*70)
    start_time = time.time()

    # Loop over each tweet id and collect the information
    with open('tweet_info.json', 'w') as file:
        print('Pulling json data for the first 200 tweets...')
        for tweet_id in tweet_ids:
            if (success % 200 ==0) and (counter > 0):
                print('\033[94m'+'\033[1m'+'Sub-task Complete!'+'\033[0m')
                print('Successful pulls: {} || failed pulls: {} || Pulls pending: {}'.format(success, failure, tweet_ids.size - counter))
                print('\nPulling json data for the next 200 tweets...')
            try:
                tweet_info = api.get_status(tweet_id, tweet_mode='extended')
                json.dump(tweet_info._json, file)
                file.write('\n')
                success+=1
            except Exception as e:
                failed_attempts[tweet_id] = e
                failure+=1
                pass
            finally:
                counter+=1

    # Print feedback on execution process  
    duration = (time.time() - start_time)/60
    failed = len(failed_attempts.keys())
    print('\033[1m' + '\033[94m' +'Task Completed!\n'+'\033[0m' + '-'*70)
    print('\033[1m'+'DISPLAYING RUNTIME SUMMARY'+'\033[0m')
    print('The entire process took: {} minutes'.format(round(duration, 2)))

    if (failed > 0):
        print('\033[91m'+'\033[1m'+'Could not pull information for '+ str(failed) + ' tweet ids:'+'\033[0m')
        print(pd.Series(failed_attempts))
    else:
        print('\033[94m'+'\033[1m'+'No failed attempts'+'\033[0m')

In [8]:
# Extract the information we want from the json file
json_tweet_details = []

with open('tweet_info.json', 'r', encoding='UTF-8') as file:
    for line in file:
        json_text = json.loads(line)
        # Extract the tweet_id, likes and retweet count
        tweet_id = json_text['id_str']
        retweets = json_text['retweet_count']
        likes = json_text['favorite_count']
        # Extract the hashtag from the json file
        hashtags_info = json_text['entities']['hashtags']
        if len(hashtags_info) !=0:
            hashtags = ['#'+item['text'] for item in hashtags_info]
        else:
            hashtags = 'None'
        # Assign these values into our list
        json_tweet_details.append({
            'tweet_id': tweet_id,
            'hashtag': hashtags,
            'retweets': retweets,
            'likes': likes}
        )
    
    json_tweet_info = pd.DataFrame(json_tweet_details)

In [9]:
json_tweet_info.head(3)

tweet_id hashtag  retweets  likes
0  892420643555336193    None      7024  33866
1  892177421306343426    None      5305  29364
2  891815181378084864    None      3488  22089

## Assessing Data
---
### A. Visual Assessment
**1. Examining a sample of 50 records from the `wrd_archive` dataframe in Jupyter notebook, including additional visual assessments in google sheets:**

In [10]:
wrd_archive.sample(50)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1199  716791146589110272                    NaN                  NaN   
133   866720684873056260                    NaN                  NaN   
819   770655142660169732                    NaN                  NaN   
198   854365224396361728                    NaN                  NaN   
1778  677895101218201600                    NaN                  NaN   
1258  710283270106132480                    NaN                  NaN   
2249  667861340749471744                    NaN                  NaN   
1437  697242256848379904                    NaN                  NaN   
1451  695794761660297217                    NaN                  NaN   
1589  686683045143953408                    NaN                  NaN   
1069  740365076218183684                    NaN                  NaN   
1081  738885046782832640                    NaN                  NaN   
598   798686750113755136                    NaN                  NaN   
1379  701545186879471618                    NaN                  NaN   
422   821813639212650496                    NaN                  NaN   
637   793286476301799424                    NaN                  NaN   
900   758740312047005698                    NaN                  NaN   
673   789628658055020548                    NaN                  NaN   
1124  730211855403241472                    NaN                  NaN   
1898  674690135443775488                    NaN                  NaN   
82    876838120628539392                    NaN                  NaN   
705   785639753186217984                    NaN                  NaN   
1367  702598099714314240                    NaN                  NaN   
406   823719002937630720                    NaN                  NaN   
1747  679111216690831360                    NaN                  NaN   
2026  671879137494245376                    NaN                  NaN   
587   799757965289017345                    NaN                  NaN   
334   832682457690300417                    NaN                  NaN   
139   865359393868664832                    NaN                  NaN   
1427  697630435728322560                    NaN                  NaN   
240   846514051647705089                    NaN                  NaN   
1995  672594978741354496                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
600   798673117451325440                    NaN                  NaN   
492   813217897535406080                    NaN                  NaN   
1368  702539513671897089                    NaN                  NaN   
1036  744971049620602880                    NaN                  NaN   
765   777885040357281792                    NaN                  NaN   
1680  682003177596559360                    NaN                  NaN   
416   822610361945911296                    NaN                  NaN   
2056  671357843010908160                    NaN                  NaN   
1053  742465774154047488                    NaN                  NaN   
414   822859134160621569                    NaN                  NaN   
954   751793661361422336                    NaN                  NaN   
59    880872448815771648                    NaN                  NaN   
189   855860136149123072           8.558585e+17           13615722.0   
1795  677314812125323265                    NaN                  NaN   
364   829141528400556032                    NaN                  NaN   
362   829449946868879360                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   

                      timestamp  \
1199  2016-04-04 00:55:01 +0000   
133   2017-05-22 18:21:28 +0000   
819   2016-08-30 16:11:18 +0000   
198   2017-04-18 16:05:17 +0000   
1778  2015-12-18 16:56:01 +0000   
1258  2016-03-17 01:55:02 +0000   
2249  2015-11-21 00:25:26 +0000   
1437  2016-02-10 02:14:42 +0000   
1451  2016-02-06 02:22:53 +0

**Notes:**
> **Quality Issues**
>- Some records appear to be replies or retweets to previously created tweets; some may contain ratings, but they are not the original tweets. This information can be observed in the `in_reply_to_status_id`,	`in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp` columns.
>- Unexpected ratings in the `rating_numerator` and `rating_denominator` columns. Examples are rating numerators as high as `666` and denominators as low as `0`.
>- Unusual dog names such as `a`, `an` and `not` in the `name` column.

> **Tidiness Issues**
>- The various stages of dog life: `doggo`, `pupper`, `puppo`, and `floofer` should be contained in one column.
>- Long and unneccessary links in the `source` column. All we need is the type of device users are tweeting from.
<br>

**2. Examining a sample of 50 records from the `img_predictions` dataframe in Jupyter notebook, including additional visual assessments in google sheets:**

In [11]:
img_predictions.sample(50)

tweet_id                                            jpg_url  \
390   673363615379013632    https://pbs.twimg.com/media/CVhEoq4WcAE8pBm.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1650  809220051211603969    https://pbs.twimg.com/media/CzrtWDbWEAAmIhy.jpg   
1522  788070120937619456    https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
53    667012601033924608    https://pbs.twimg.com/media/CUG0bC0U8AAw2su.jpg   
793   690735892932222976    https://pbs.twimg.com/media/CZX8nyeVAAEstKM.jpg   
420   674042553264685056    https://pbs.twimg.com/media/CVquIDRW4AEJrPk.jpg   
1833  836753516572119041    https://pbs.twimg.com/media/C5y-4VwWcAIcaoj.jpg   
1253  748307329658011649    https://pbs.twimg.com/media/CmKFi-FXEAAeI37.jpg   
1770  827600520311402496    https://pbs.twimg.com/media/C3w6RYbWQAAEQ25.jpg   
582   678798276842360832    https://pbs.twimg.com/media/CWuTbAKUsAAvZHh.jpg   
504   675888385639251968    https://pbs.twimg.com/media/CWE85snWIAEG5ES.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
466   675015141583413248    https://pbs.twimg.com/media/CV4iqh5WcAEV1E6.jpg   
630   680934982542561280    https://pbs.twimg.com/media/CXMqwIQWcAA2iE0.jpg   
985   707610948723478529    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1547  792773781206999040    https://pbs.twimg.com/media/CwB_i-zXEAEiP29.jpg   
901   700002074055016451    https://pbs.twimg.com/media/CbboKP4WIAAw8xq.jpg   
315   671735591348891648    https://pbs.twimg.com/media/CVJ79MzW4AEpTom.jpg   
71    667200525029539841    https://pbs.twimg.com/media/CUJfVMPXIAAgbue.jpg   
1734  821522889702862852    https://pbs.twimg.com/media/C2aitIUXAAAG-Wi.jpg   
203   669749430875258880    https://pbs.twimg.com/media/CUttjYtWcAAdPgI.jpg   
219   670079681849372674    https://pbs.twimg.com/media/CUyZ6mVW4AI8YWZ.jpg   
824   693231807727280129    https://pbs.twimg.com/media/CZ7aplIUsAAq-8s.jpg   
1073  716802964044845056    https://pbs.twimg.com/media/CfKYfeBXIAAopp2.jpg   
1849  839990271299457024    https://pbs.twimg.com/media/C6g-sX-VsAAHfJ9.jpg   
1501  784431430411685888    https://pbs.twimg.com/media/CuLcNkCXgAEIwK2.jpg   
1392  767122157629476866    https://pbs.twimg.com/media/CqVdiBJWIAEDZB4.jpg   
1974  869772420881756160    https://pbs.twimg.com/media/DBINZcxXgAQ-R6P.jpg   
446   674664755118911488    https://pbs.twimg.com/media/CVzkA7-WsAAcXz6.jpg   
1900  851464819735769094    https://pbs.twimg.com/media/C9ECujZXsAAPCSM.jpg   
528   676776431406465024  https://pbs.twimg.com/ext_tw_video_thumb/67677...   
1692  815990720817401858    https://pbs.twimg.com/media/C1L7OVVWQAIQ6Tt.jpg   
45    666786068205871104    https://pbs.twimg.com/media/CUDmZIkWcAAIPPe.jpg   
562   677918531514703872    https://pbs.twimg.com/media/CWhzTbzWUAAEAUN.jpg   
1107  723673163800948736    https://pbs.twimg.com/media/CgsA5eFWgAAu0qn.jpg   
1606  800459316964663297    https://pbs.twimg.com/media/CxvNfrhWQAA2hKM.jpg   
733   686749460672679938    https://pbs.twimg.com/media/CYfS75fWAAAllde.jpg   
1623  803638050916102144  https://pbs.twimg.com/ext_tw_video_thumb/80363...   
1492  782969140009107456    https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg   
1236  746726898085036033    https://pbs.twimg.com/media/ClzoJz7WYAELHSf.jpg   
1632  805826884734976000  https://pbs.twimg.com/ext_tw_video_thumb/80582...   
163   668960084974809088    https://pbs.twimg.com/media/CUifpn4WUAAS5X3.jpg   
196   669597912108789760    https://pbs.twimg.com/media/CUrjvxiVEAA94dH.jpg   
33    666430724426358785    https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
358   672614745925664768    https://pbs.twimg.com/media/CVWbitUW4AAzclx.jpg   
227   670374371102445568    https://pbs.twimg.com/media/CU2l7yvXAAUyYIJ.jpg   
1707  817777686764523521  https://pbs.twimg.com/ext_tw_video_thumb/81777...   
559   677700003327029250    https://pbs.twimg.com/media/CWesj06W4AAIKl8.jpg   
1476  780496263422808064    https://pbs.twimg.com/media/Ck2d7tJWUA

**Notes:**
> **Quality Issues**
>- The Prediction in columns `p1`, `p2` and `p3` are not uniformly formatted. Some names are lowercase, some are uppercase and some are titlecase.
>- The predictions above also have words seperated by underscores instead of spaces.

> **Tidiness Issues**
>- From `p1`, `p2` and `p3`, we only need the most confident prediction that corresponds to an actual dog breed.
<br>

**3. Examining a sample of 50 records each from the `json_tweet_info` dataframe in Jupyter notebook**

In [12]:
json_tweet_info.sample(50, random_state=1)

tweet_id       hashtag  retweets  likes
579   798682547630837760          None      4442      0
1255  708479650088034305          None       622   2352
47    882992080364220416          None      3250  21009
1117  727155742655025152          None      1202   3340
614   793614319594401792          None      2937      0
108   870804317367881728          None      5293  29850
6     890971913173991426          None      1673  10382
2162  668892474547511297          None       126    354
1178  715733265223708672          None      1540   4304
2254  667182792070062081          None      5291  12569
1687  680191257256136705          None       445   1985
576   798697898615730177          None      6114      0
644   790723298204217344          None      2154      0
17    888804989199671297          None      3545  22501
2093  670361874861563904          None        56    290
1412  696754882863349760          None       329   1374
403   821886076407029760          None      2149  10697
1053  738883359779196928          None       725   3056
65    879415818425184262          None     37510  92986
1653  681694085539872773          None      3701  11849
1731  678643457146150913          None       361   1881
1264  708026248782585858          None      1758   4079
632   792394556390137856          None      4026  12752
464   815390420867969024          None      3578   9841
1105  728653952833728512          None       953   3069
304   834458053273591808          None      1556   9050
1558  686730991906516992          None      1096   3817
486   813051746834595840          None      6876  19953
1232  710153181850935296          None       816   2695
1950  672968025906282496          None       474   1189
1378  699079609774645248          None       579   2267
828   765371061932261376          None      2010   6690
1592  684914660081053696          None      1327   3231
1156  718613305783398402          None       431   2265
126   867051520902168576          None      6799  28749
1272  707629649552134146          None       788   2363
1511  689623661272240129          None       583   2046
1577  685663452032069632          None      1338   2983
56    881268444196462592  [#Canada150]      4503  20509
1436  694342028726001664          None       433   1468
894   756275833623502848          None      1410   5990
757   775898661951791106          None     15576      0
1525  688894073864884227          None       623   2066
368   826615380357632002          None      3695      0
321   832397543355072512          None      2042  11269
231   846153765933735936          None      8203  29186
120   868622495443632128          None      4486  23668
2323  666020888022790149          None       424   2292
1732  678424312106393600          None      2291   5005
1551  687124485711986689          None       464   2026

**Notes:**
> **Quality Issue**
>- It seems that where hashtags are present, they are stored in list format rather than as individual strings.

### B. Programmatic Assessment
#### 1. WRD Archive Data

In [13]:
wrd_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

**Notes**
>- `tweet_id` stored as `int` instead of `string/object` type.
>- `181` records are retweets and `78` records are replies. We won't be using these records in our analysis.
>- `timestamp` column is stored as `string/object` type rather than as a `datetime` object.
>- The `expanded_urls` column has some null records

Lets narrow our observations to only records where the `expanded_url` entries are null:

In [14]:
print('Collating the number of null entries for records with missing expanded urls..')
print(wrd_archive[wrd_archive['expanded_urls'].isnull()].isnull().sum())

Collating the number of null entries for records with missing expanded urls..
tweet_id                       0
in_reply_to_status_id          4
in_reply_to_user_id            4
timestamp                      0
source                         0
text                           0
retweeted_status_id           58
retweeted_status_user_id      58
retweeted_status_timestamp    58
expanded_urls                 59
rating_numerator               0
rating_denominator             0
name                           0
doggo                          0
floofer                        0
pupper                         0
puppo                          0
dtype: int64


>Tweets with missing `expanded_urls` are either retweets or likes. We won't be including these records in our analysis.

In [15]:
# Check the archive for duplicate records
duplicates = wrd_archive.duplicated().sum()
print('wrd_archive has {} number of duplicate records'.format(duplicates))

wrd_archive has 0 number of duplicate records


In [16]:
# Examine the unique values in the source column
print(Color.bold+'Examining unique values in the source column'+Color.end)
for i, item in enumerate(wrd_archive['source'].unique()):
    print(i, ': ', item)

Examining unique values in the source column
0 :  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
1 :  <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>
2 :  <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>
3 :  <a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>


>- We only want to extract the information between the `<a> </a>` tags, signalling the tweet source.

In [17]:
# Examine the text column and expanded_url columns
pd.set_option("display.max_colwidth",0)
wrd_archive[['text', 'expanded_urls']].sample(5)

text  \
1124  This is Ziva. She doesn't know how her collar works. 11/10 would totally fix for her https://t.co/K7pthJXjWE                            
281   RT @alexmartindawg: THE DRINK IS DR. PUPPER 10/10 good pun @matt___nelson @GoodDogsGame https://t.co/act3duiqbL                         
415   RT @dog_rates: This is Paisley. She really wanted to be president this time. Dreams officially crushed. 13/10 https://t.co/liJGwMp17E   
1510  This is Bailey. She likes flowers. 12/10 https://t.co/YBENhr24FV                                                                        
30    @NonWhiteHat @MayhewMayhem omg hello tanner you are a scary good boy 12/10 would pet with extreme caution                               

                                                                                                                                                                                                                                                                                                                                                                                        expanded_urls  
1124  https://twitter.com/dog_rates/status/730211855403241472/photo/1                                                                                                                                                                                                                                                                                                                                  
281   https://twitter.com/alexmartindawg/status/839289919298224128/photo/1,https://twitter.com/alexmartindawg/status/839289919298224128/photo/1,https://twitter.com/alexmartindawg/status/839289919298224128/photo/1,https://twitter.com/alexmartindawg/status/839289919298224128/photo/1                                                                                                              
415   https://twitter.com/dog_rates/status/822489057087389700/photo/1,https://twitter.com/dog_rates/status/822489057087389700/photo/1,https://twitter.com/dog_rates/status/822489057087389700/photo/1,https://twitter.com/dog_rates/status/822489057087389700/photo/1,https://twitter.com/dog_rates/status/822489057087389700/photo/1,https://twitter.com/dog_rates/status/822489057087389700/photo/1  
1510  https://twitter.com/dog_rates/status/691444869282295808/photo/1,https://twitter.com/dog_rates/status/691444869282295808/photo/1                                                                                                                                                                                                                                                                  
30    NaN

On clicking the links and testing them, one would discover that both the `text` and `expanded url` links lead to the same tweet for each record. Some records also have multiple expanded urls seperated by commas, all leading to the same tweet. As a result, we can make the following notes:
>- The `text` column contains both the tweet text and tweet url. 
>- The same tweet url is already present in the `expanded_urls` column

In [18]:
# Examine the distribution of ratings in the dataset
wrd_archive[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count  2356.000000       2356.000000       
mean   13.126486         10.455433         
std    45.876648         6.745237          
min    0.000000          0.000000          
25%    10.000000         10.000000         
50%    11.000000         10.000000         
75%    12.000000         10.000000         
max    1776.000000       170.000000

In [19]:
# Examine the unique values in rating numerator and denominator
print(Color.bold+'Unique rating numerators'+Color.end)
print(wrd_archive['rating_numerator'].unique())
print(Color.bold+'\nUnique rating denominators'+Color.end)
print(wrd_archive['rating_denominator'].unique())

Unique rating numerators
[  13   12   14    5   17   11   10  420  666    6   15  182  960    0
   75    7   84    9   24    8    1   27    3    4  165 1776  204   50
   99   80   45   60   44  143  121   20   26    2  144   88]

Unique rating denominators
[ 10   0  15  70   7  11 150 170  20  50  90  80  40 130 110  16 120   2]


Though WeRateDogs post can have numerators higher than 10, they almost always have denominators of 10. Having numerators as high as 1776 and denominators as low as 0 prompts us to inspect the dataframe further:

In [20]:
# Assess instances where rating numerators > 20 and denominators are !=10
rating_check_df = (wrd_archive[(wrd_archive['rating_numerator'] > 20) | (wrd_archive['rating_denominator']!=10)])
# filter out the retweets
rating_check_df = (rating_check_df[rating_check_df['retweeted_status_id'].isnull()])
# filter out the replies
rating_check_df = (rating_check_df[rating_check_df['in_reply_to_status_id'].isnull()])
# Finally examine the text and the ratings
print('{} records found'.format(rating_check_df.shape[0]))
rating_check_df[['text', 'rating_numerator', 'rating_denominator']]

22 records found


text  \
433   The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                                         
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
695   This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                            
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq           
902   Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                                              
979   This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh                                                                                              
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ                                
1120  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                                                   
1165  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                                            
1202  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                                       
1228  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                                               
1254  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                                                
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK                          
1351  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                                          
1433  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                                                
1635  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55                                
1662  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5                                 
1712  Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD                                                                  
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                                                      
1843  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                                                 
2074  After so many requests... here you go.\n\nGood dogg. 420/10 https://t.co/yfAAo1gdeY                                                                                         
2335  This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv                                    

      rating_numerator  rating_denominator  
433   84

>- Some ratings were erroneously pulled from the original tweet. Especially when dates (e.g 24/7 and 9/11) or decimal ratings (e.g 11.27/10 and 9.75/10) are included in the tweet text.
>- Overly high ratings appear to be addressed to groups of dogs as can be seen in the following tweet urls among others: [165/150](https://t.co/HmwrdfEfUE), [84/70](https://t.co/NIYC820tmd), [88/80](https://t.co/y93p6FLvVw).

In [21]:
# Examine the name column further especially names with 4 characters or less
print(Color.bold+'Specially examine names with four string characters or less..'+Color.end)
print(wrd_archive.name[wrd_archive.name.apply(lambda x: len(x)<=4)].unique())

Specially examine names with four string characters or less..
['None' 'Jax' 'Zoey' 'Koda' 'Ted' 'Jim' 'Zeke' 'such' 'Maya' 'Earl' 'Lola'
 'Yogi' 'Noah' 'Gus' 'Alfy' 'Koko' 'Rey' 'Gary' 'a' 'Jack' 'Emmy' 'Beau'
 'Aja' 'Cash' 'Coco' 'Jed' 'Kody' 'Dawn' 'Cody' 'Lili' 'Dave' 'Burt'
 'Carl' 'Thor' 'Luna' 'Arya' 'Iggy' 'Kyle' 'Leo' 'Odin' 'Tuck' 'Hank'
 'Ken' 'Max' 'Odie' 'Arlo' 'Lucy' 'Ava' 'Rory' 'Eli' 'Ash' 'Tobi' 'not'
 'Kuyu' 'Pete' 'Kyro' 'Loki' 'Mia' 'one' 'Mutt' 'Bear' 'Kona' 'Phil' 'Ike'
 'Mo' 'Toby' 'Nala' 'Gabe' 'Luca' 'Finn' 'Anna' 'Bo' 'Tom' 'Dido' 'Levi'
 'Alf' 'Sky' 'Tyr' 'Mary' 'Moe' 'Halo' 'Sam' 'Ito' 'Milo' 'Cali' 'Duke'
 'Chef' 'Doc' 'Sobe' 'Iroh' 'Ruby' 'Mack' 'Juno' 'Lily' 'Newt' 'Nida'
 'BeBe' 'mad' 'Dale' 'Hero' 'Godi' 'Dash' 'Bell' 'Jay' 'Mya' 'an' 'Huck'
 'very' 'O' 'Blue' 'Fizz' 'Chip' 'Grey' 'Al' 'just' 'Lou' 'Tito' 'Brat'
 'Tove' 'my' 'Kota' 'Eve' 'Rose' 'Theo' 'Fido' 'Emma' 'Gert' 'Dex' 'Ace'
 'Fred' 'Zoe' 'Blu' 'his' 'Cora' 'Abby' 'Geno' 'Beya' 'Kilo' 'Doug' 'Aq

>- Again we notice more unusual names like `the`, `my`, `by`, `his`, `all`, `mad`, `life`, `very`, `old`, `this`, `just` etc. All these unusual names are formatted in lower case, while the more viable names are properly capitalized.

In [22]:
#Examine the dog stage columns
for dog_stage in wrd_archive.columns[-4:]:
    print(Color.bold+'\nValue counts for {} column'.format(dog_stage)+Color.end)
    print(wrd_archive[dog_stage].value_counts())


Value counts for doggo column
None     2259
doggo    97  
Name: doggo, dtype: int64

Value counts for floofer column
None       2346
floofer    10  
Name: floofer, dtype: int64

Value counts for pupper column
None      2099
pupper    257 
Name: pupper, dtype: int64

Value counts for puppo column
None     2326
puppo    30  
Name: puppo, dtype: int64


>- Asides the fact that we have to tidy up these columns into one, everything looks good.
#### 2. Image Predictions Data